[settings] support 'single_select_page_with_search' field type

chihsuan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

This PR implements support for the `single_select_page_with_search` field type in the settings editor, addressing part of issue #54592. The implementation:

* Introduces `SingleSelectPageWithSearchEdit` component for enhanced page selection with search functionality based on the existing `__experimentalSelectControl` component from `@woocommerce/components`
* Integrates with WordPress core-data store for efficient page data management


### Screenshots or screen recordings:


https://github.com/user-attachments/assets/6db3c261-6072-4644-b15f-0947ccc0a8e6


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:


1. Enable `settings` feature flag via WCA tester
2. Go to `/wp-admin/admin.php?page=wc-settings&tab=advanced`
3. Confirm that `Cart page`, `Checkout page`, `My account page`, and `Terms and conditions page` fields are present
4. Verify the following functionality:
   * Search input filters pages correctly
   * Loading spinner appears while fetching pages
   * Remove selection button (X) appears when a page is selected, and clicking it removes the selection
   * Selected page persists after save


<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [x] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

Support 'single_select_page_with_search' field type for setting editor

</details>

chihsuan committed:
Update dependencies in pnpm-lock.yaml and package.json for WooCommerce components

- Updated the version of `@babel/core` in `ts-jest` to 7.26.0.
- Added `@woocommerce/components` as a workspace dependency in `settings-editor/package.json`.
- Updated asset dependencies in `Init.php` to include `wc-components` alongside `wp-components` for improved asset management.

chihsuan committed:
Add SingleSelectPageWithSearchEdit component and SelectSuffix for enhanced page selection

- Introduced `SingleSelectPageWithSearchEdit` component for selecting pages with search functionality.
- Implemented `SelectSuffix` component to display loading state and remove selection.
- Integrated hooks for fetching and filtering pages, improving user experience in selecting pages.
- Added utility functions for formatting and managing selected page items.

chihsuan committed:
Add hooks and utility functions for page selection in settings editor

- Introduced `usePageSearch` and `useSelectedItem` hooks to manage page selection and search functionality.
- Added utility functions for formatting page items, enhancing the user experience in selecting pages.
- Created `Suffix` component to display loading states and allow removal of selected pages.
- Updated `SingleSelectPageWithSearchEdit` component to utilize new hooks and components for improved functionality.

moon0326 reviewed (commented):
LGTM and tested well. I have one comment on the search field behavior.

I find it a bit confusing that I have to manually delete "Cart (ID: 7)" before the "Search for a page..." input appears.

https://github.com/user-attachments/assets/037e5a05-082b-483b-abcf-001fa2155fbd

Would it make sense to show "Search for a page..." immediately on click? Then, we could re-display "Cart (ID: 7)" if focus is lost and nothing was selected.






The legacy settings page handled this more clearly—it's immediately obvious what actions are possible.

![Screenshot 2025-03-31 at 12 43 23 PM](https://github.com/user-attachments/assets/6f1cb2d9-b5d8-4325-be67-a6ef70b64eba)

chihsuan committed:
Add onBlur prop to SelectControl and enhance SingleSelectPageWithSearchEdit component

- Introduced onBlur prop in SelectControl to handle blur events and pass input value.
- Updated SingleSelectPageWithSearchEdit to manage searching state with onFocus and onBlur handlers.
- Improved user experience by conditionally rendering selected items based on search state.

chihsuan commented:
> Would it make sense to show "Search for a page..." immediately on click? Then, we could re-display "Cart (ID: 7)" if focus is lost and nothing was selected.

Thanks for the feedback! @moon0326 That's a good suggestion. I've updated it to behave as you described in fbd7c39e6f8a78c29c0318c95840eef0d6c915df.

https://github.com/user-attachments/assets/7ab6ab68-b7ff-4e42-8bc7-b4520e2fcc80

moon0326 reviewed (approved):
Thanks for the improvement! LGTM and tested well 👍 🚀 
___
#### Review checklist
| Parameter  | Result |
| ------------- | ------------- |
| Approach  | ✅ All good |
| Implementation  | ✅ All good |
| Automated tests  | Not applicable |
| Naming  | ✅ All good |
| Typing  |   ✅ All good |
| Code comments  | Not applicable |
| Changelog  |   ✅ All good |
| Documentation  | ✅ All good |
| PR description  | ✅ All good |
| Testing instructions  | ✅ All good |
| Manual testing  | ✅ Works as expected | 
 
 
#### Scenarios tested
| Scenario  | Result | Screenshot |
| ------------- | ------------- | ------------- |
| select component loads with selected item  | ✅ Functionality worked correctly | ![Screenshot 2025-04-01 at 9 31 32 AM](https://github.com/user-attachments/assets/21c09114-65eb-422c-bec9-37cd54d009fd)|
| Search works as expected | ✅ Functionality worked correctly | ![Screenshot 2025-04-01 at 9 31 39 AM](https://github.com/user-attachments/assets/bdde516c-e90c-4082-8eb7-b21bc4d4208f) |

psealock commented on packages/js/settings-editor/src/components/single-select-page-with-search-edit/index.tsx:
```diff
@@ -0,0 +1,149 @@
+										getItemProps={ getItemProps }
+									>
+										{ item.label }
+									</MenuItem>
```
Not sure if this should be at the component level, but cursor should reflect a button when hovering over a menu item.

<img width="393" alt="image" src="https://github.com/user-attachments/assets/26ebf914-67df-4539-a31d-fd76187a9abe" />

chihsuan replied:
Good point! I think this should be handled at the component level. 

I observed that the select controls in the product editor do not display the pointer cursor on hover, similar to the site editor. I don't know if it's intentional.  🤔  I'll check with the design team.

![Screenshot 2025-04-02 at 11 20 53](https://github.com/user-attachments/assets/9f10009a-7181-43eb-9565-b67759b49fe6)

psealock commented on packages/js/settings-editor/src/components/single-select-page-with-search-edit/index.tsx:
```diff
@@ -0,0 +1,149 @@
+				items={ searchedItems }
+				selected={ isSearching ? null : selectedItem }
+				onSelect={ handleSelect }
+				onFocus={ () => setIsSearching( true ) }
```
I would expect to fire a request at this point with search term empty string so that all pages are returned. What do you think?

chihsuan replied:
That sounds like a better user experience. But I see that the legacy settings input only triggers the request when the user types at least 3 characters. Do you happen to know why?

psealock commented on packages/js/settings-editor/src/components/single-select-page-with-search-edit/index.tsx:
```diff
@@ -0,0 +1,149 @@
+				selected={ isSearching ? null : selectedItem }
+				onSelect={ handleSelect }
+				onFocus={ () => setIsSearching( true ) }
+				onBlur={ () => setIsSearching( false ) }
```
When tabbing through the inputs, this component isn't properly blurring. It still has blur styles and the current value doesn't return.

<img width="1312" alt="image" src="https://github.com/user-attachments/assets/38055c63-29b7-426d-96af-fd730ca825a1" />

chihsuan replied:
Good catch! I'll take a look.

chihsuan replied:
Fixed in 71926099e5e52f26fd5366e746fd53fe06526304

https://github.com/user-attachments/assets/19503aef-a274-4d75-9cc0-6dac2fba46d4

psealock commented on packages/js/settings-editor/src/components/single-select-page-with-search-edit/index.tsx:
```diff
@@ -0,0 +1,149 @@
+						<Icon icon={ search } size={ 20 } />
+						<Menu isOpen={ isOpen } getMenuProps={ getMenuProps }>
+							{ isFetching ? (
+								<Spinner />
```
Some padding on this spinner would be good.

chihsuan replied:
I've added styles for the spinner. It's the same as the one in the product editor. cfb847c8cae97f3867b7b6bc9dda2961cb36f2eb

chihsuan replied:
https://github.com/user-attachments/assets/62369231-4332-4bc6-aa09-5664c8ccab80

psealock commented on packages/js/settings-editor/src/components/single-select-page-with-search-edit/hooks.ts:
```diff
@@ -0,0 +1,122 @@
+ * @param value - The value of the selected page item.
+ * @return The selected page item and a boolean indicating if the page is loading.
+ */
+export const useSelectedItem = ( value: string ) => {
```
This should probably be `useSelectedPage` considering the return values. I think a page becomes an item only after its been run through `formatPageToItem`.

chihsuan replied:
I think `useSelectedItem` is the correct name here. The hook actually returns the selected item, not the selected page. 😅

psealock commented on packages/js/settings-editor/src/components/single-select-page-with-search-edit/hooks.ts:
```diff
@@ -0,0 +1,122 @@
+ * @param exclude      - The pages to exclude from the search results.
+ * @return The searched pages and a boolean indicating if the pages are loading.
+ */
+export const usePageSearch = (
```
Same page v. item naming here. This function accepts a selected item, so should be useItemSearch... am I following correctly?

chihsuan replied:
You're right. I'll update the name.

chihsuan replied:
Renamed `usePageSearch` hook to `useItemSearch` in 7d0b79ec0330036adfaaaf155115152b35b87bf9.

psealock reviewed (commented):
Nice work here @chihsuan, this is a complicated component. There's a few things off in the interaction and some naming things to address.

chihsuan committed:
Fix SingleSelectPageWithSearch component focus state

- Replaced searching state with focus state in SingleSelectPageWithSearch component.
- Updated handlers for onFocus and onBlur to reflect the new focus management.
- Adjusted Suffix component to conditionally render based on focus state, improving user experience during selection.

chihsuan committed:
Rename usePageSearch hook to useItemSearch for clarity in searching page items. Update related documentation and references in the SingleSelectPageWithSearch component to reflect the new naming convention.

chihsuan committed:
Add form controls styles and update SingleSelectPageWithSearch component

- Introduced new styles for form controls in a dedicated SCSS file.
- Updated SingleSelectPageWithSearch component to wrap the loading spinner in a new div for better styling and layout management.

chihsuan committed:
Merge branch 'trunk' into add/support-single-select-page-with-search

chihsuan committed:
Refactor page selection hooks and update styles

- Renamed `useItemSearch` to `useItems` for clarity and updated its implementation to fetch all page items.
- Modified `SingleSelectPageWithSearch` component to utilize the new `useItems` hook and adjusted related logic.
- Added cursor pointer style to select control popover menu items for better user interaction.

chihsuan committed:
Merge branch 'trunk' into add/support-single-select-page-with-search

chihsuan committed:
Merge branch 'trunk' into add/support-single-select-page-with-search

chihsuan committed:
Fix lint

chihsuan commented:
Hey @woocommerce/flux Could I get a review or approval for the PR to merge it? 

The only changes related to monorepo are adding `@woocommerce/components` as a dependency for settings-editor.

prettyboymp reviewed (approved):
No concerns here.
